connectors/spark/src/test/java/io/unitycatalog/spark/auth/AwsCredentialRenewalTest.java

```diff
@@ -0,0 +1,362 @@
+    //    filesystem credentials are renewed the expected number of times.
+    // 3. Collect the credential renewal count.
+    //
+    // The core logic resides in the mapFunction. We use the Delta table’s Hadoop configuration
+    // to initialize the filesystem, simulating Delta table operations within a Spark executor.
+    // This allows us to accurately track how many times credentials are renewed within a task.
+    //
+    // It is possible for a Spark job to create multiple independent filesystem instances,
+    // which may misleadingly appear to renew credentials correctly even when they do not.
+    // We adopt this simulation approach because directly accessing a Spark task’s internal
+    // filesystem instance to measure credential renewals is not feasible.
+    List<Row> rows =
+        session
+            .read()
+            .format("delta")
+            .table(TABLE_NAME)
+            .toJavaRDD()
+            .map(
+                row -> {
+                  Configuration conf = serialConf.value();
+                  S3CredFileSystem fs = (S3CredFileSystem) FileSystem.get(new URI(location), conf);
+
+                  for (int refreshIndex = 0; refreshIndex < 10; refreshIndex += 1) {
+                    // Pre-check before the credential renewal.
+                    fs.getFileStatus(locPath);
+                    assertThat(fs.renewalCount()).isEqualTo(refreshIndex);
+
+                    // Advance the clock to trigger the renewal.
+                    testClock().advance(Duration.ofMillis(DEFAULT_INTERVAL_MILLIS));
+
+                    // Post-check after the credential renewal.
+                    fs.getFileStatus(locPath);
+                    assertThat(fs.renewalCount()).isEqualTo(refreshIndex + 1);
+                  }
+
+                  return RowFactory.create(10);
+                })
+            .collect();
+
+    assertThat(rows.stream().map(r -> r.getInt(0)).collect(Collectors.toList()))
+        .isEqualTo(ImmutableList.of(10));
+  }
+
+  @Test
+  public void testDeltaReadWriteRenewal() throws Exception {
+    String location = String.format("s3://%s%s/delta", BUCKET_NAME, dataDir.getCanonicalPath());
+    StructType schema =
+        new StructType(
+            new StructField[] {
+              new StructField("id", DataTypes.IntegerType, false, Metadata.empty()),
+              new StructField("partition", DataTypes.IntegerType, false, Metadata.empty()),
+            });
+
+    // Create the external delta path-based table.
+    S3CredFileSystem.credentialCheckEnabled = false;
+    sql("CREATE TABLE delta.`%s` (id INT) USING delta PARTITIONED BY (partition INT)", location);
+    S3CredFileSystem.credentialCheckEnabled = true;
+
+    // Create the catalog table referring to external table.
+    sql("CREATE TABLE %s USING delta LOCATION '%s'", TABLE_NAME, location);
+
+    // Insert 4 rows into each partition where (id % 3) equals the partition key.
```

Comment:
why three INSERT with `INSERT INTO ... VALUES (1, 1), (1, 2), (1,3)`
that will create 3 files with 1 row each because partition column has 3 values.

Reply:
Em, let me simply this test.  Done !